src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -138,6 +141,20 @@ public void run() {
                 // Clear the thread locals
                 mlThread.eraseThreadLocals();
             }
+            if (blockForCheckpoint) {
+                try {
+                    synchronized (this) {
+                        waitingForCheckpoint = true;
+                        notify();
+                        while (blockForCheckpoint) {
+                            wait();
+                        }
```

Comment:
Once we've got here, is it possible to ensure Cleaners has been called, and drop separate registration of Cleaners?

A concurrent cleaner registration is not a problem, as that depends on GC which is not predictable. I.e. if that happens, a slight race may also cause the cleaner to run after restore.

Reply:
I don't follow. PhantomCleanableRefs are cleaned strictly after this point (these have later priority), other cleaners will be queued up by GC but not called until restore.

Reply:
I mean, once we synchronized cleaner thread from picking up PhantomCleanableRefs, there is apparently no reason to go to the next priority and discover cleanups through registrations in the CRaC -- instead, all cleanups are discoverable thorugh the list maintained by the cleaner. It seems we can go through the list right from this Resource and run cleanups.

Reply:
Alright, you mean that PCRs won't be Resources anymore and we'll clean them up whether are these already in the `queue` or not, without waiting for the GC to move them into that.

Reply:
Something like that, but indeed the interaction between RefQueue and the list maintained by the cleaner should reviewed in the context is it safe and correct to do that.